config/vufind/PAIA.ini

```diff
@@ -18,6 +18,19 @@ baseUrl = ""
 ; same as for DAIA cache (see cacheLifetime setting in DAIA.ini).
 ;paiaCache = false
 
+; A PAIA auth server acts as OAuth authorization server (RFC 6749) with
+; password credentials grant, as defined in section 4.3 of OAuth specification,
+; and/or client credentials grant, as defined in section 4.4 of the OAuth
+; specification.
+; Server specific grant type should be set ('password' or 'client_credentials') but
+; defaults to 'password'.
+;grantType = "password"
+
+; If grant type is set to client_credentials the client credentials must be
+; set below.
+;client_username = 'client'
+;client_password = 'password'
+
```

Comment:
I don't understand, why we need the credentials to be set here in the config. The PAIA login is always patron specific and as far as I understand it, the difference between client_credentials and password mode is the way the authentication is done (password authentication sends the username and password in the POST request, while client_credentials sends them in the HTTP header in an base64 encoded form). So, why would we need to set (global) credentials in the configuration? Shouldn't we rather use the credentials the user provided in the login form? If we cannot use those, which credentials should be provided in the config (do we need a dummy user for that?)?

Reply:
The `grant_type` `client_credential` in PAIA is not patron centered, i.e. the credentials of the patron as a client, but service centered, i.e. the credentials for the service accessing the PAIA API as a client. Therefore a shared secret is used and needs to be stored in the configuration file. Of course this will allow to the VuFind client to act as any patron without proper authentication. Therefore this method is of best use when combined with authorization via Shibboleth or equivalent.
Maybe this should be pointed out more precisely in the configuration comments?

Reply:
Is it possible that we're conflating two different things here? One issue is whether the credentials come from the end user or are hard-coded in the configuration file. The other issue is how we pass the credentials to the server -- via HTTP POST or HTTP Basic Authentication. Is it necessarily true that user-provided credentials must always be POSTed while pre-configured credentials must always be used for Basic Authentication? Would it make more sense to have two separate switches for these things? Is it possible that trying to use formal OAuth/PAIA language actually makes some things less clear in this case?

(I don't presuppose answers to any of those questions, but I think that answering some of them may prove helpful). :-)

Reply:
PAIA login (cf. http://gbv.github.io/paia/paia.html#login) supports two ways of authentication identified by the POST parameter `grant_type`:
1. `password` - athentication is done by providing patron `id` and `password` in POST data (GET is optional)
2. `client_credential` - authentication is done by providing shared secret consisting of concatenated client-username and client-password plus base64 encoding. This authenticates the client service (i.e. VuFind) and a patron identifier needs to be passed along to actually retrieve patron data or perform actions for the patron.

> Is it necessarily true that user-provided credentials must always be POSTed while pre-configured credentials must always be used for Basic Authentication?
According to the PAIA specs, yes: "The request header “Access” is required for client credentials grant with HTTP basic authentification as defined in [RFC 2617](http://tools.ietf.org/html/rfc2617)."

But both authentication methods need POST data: `grant_type`, `scope for both, `patron` only for `grant_type=client_credential` and `username`, `password` only for `grant_type=password`.

Does this make things clearer? :)

Reply:
Thanks, @lahmann, that's very helpful! I would suggest adding something like this to the .ini comments:

When grantType is set to password (the default), user-entered credentials will be passed along to the PAIA server for authentication against the API. When grantType is set to client_credentials, then the "shared secret" credentials provided in clientUsername and clientPassword below will always be used for PAIA API authentication.

I think that helps to clarify what will actually happen, in case people do not understand the specifications being referenced in the existing comments (which I think you could also retain, or not, depending on your preferences).

Reply:
Thanks @demiankatz, I added the comment and refactored the credential names. I think this should make things more understandable?

Reply:
I was thinking of putting the comment slightly higher in the file, but I think this works too. Maybe @olli-gold can comment on the clarity since he can look at it with a fresh/objective perspective. :-)

Reply:
Thanks for clarification. I understand better now, why we need the central username and password for client authentication in the config. But I think the comment is a bit too technical and would suggest to refactor it a little bit. 

Thus, I would cut off the first sentence in line 29f, as it should be clear, that you do not need to fill out this section for password mode, but only for client_credentials. And for the second sentence I would propose something like that:
When grantType is set to client_credentials, please insert the "shared secret" credentials provided from your PAIA provider in clientUsername and clientPassword below.

Reply:
Just giving you a thumb up, that looks good to me. Thank you!